src/EditorFeatures/CSharpTest/CodeActions/SyncNamespace/SyncNamespaceTests_MoveFile.cs

```diff
@@ -0,0 +1,310 @@
+            expectedFolders.Add(new[] { "B", "C", "D", "E" });
+            expectedFolders.Add(new[] { "B.C", "D", "E" });
+            expectedFolders.Add(new[] { "B", "C.D", "E" });
+
+            var documentPath1 = CreateDocumentFilePath(Array.Empty<string>(), "File1.cs");
+            var documentPath2 = CreateDocumentFilePath(new[] { "B", "C.D" }, "File2.cs");   // file2 is in <root>\B\C.D\
+            var documentPath3 = CreateDocumentFilePath(new[] { "B.C" }, "File3.cs");   // file3 is in <root>\B.C\
+            var code =
+$@"
+<Workspace>
+    <Project Language=""C#"" AssemblyName=""Assembly1"" FilePath=""{ProjectFilePath}"" DefaultNamespace=""{defaultNamespace}"" CommonReferences=""true"">
+        <Document Folders=""{documentPath1.folder}"" FilePath=""{documentPath1.filePath}""> 
+namespace [||]{declaredNamespace}
+{{    
+    class Class1
+    {{
+    }}
+}}  
+        </Document>        
+        <Document Folders=""{documentPath2.folder}"" FilePath=""{documentPath2.filePath}""> 
+namespace Foo
+{{    
+    class Class2
+    {{
+    }}
+}}  
+        </Document>      
+        <Document Folders=""{documentPath3.folder}"" FilePath=""{documentPath3.filePath}""> 
+namespace Foo
+{{    
+    class Class2
+    {{
+    }}
+}}  
+        </Document>  
+    </Project>
+</Workspace>";
+            await TestMoveFileToMatchNamespace(code, expectedFolders);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CodeActionsSyncNamespace)]
+        public async Task MoveFile_FromOneFolderToAnother1()
+        {
+            var defaultNamespace = "A";
+            var declaredNamespace = "A.B.C.D.E";
+
+            var expectedFolders = new List<string[]>();
+            expectedFolders.Add(new[] { "B", "C", "D", "E" });
+            expectedFolders.Add(new[] { "B.C", "D", "E" });
+
+            var documentPath1 = CreateDocumentFilePath(new[] { "B.C" }, "File1.cs");                          // file1 is in <root>\B.C\
+            var documentPath2 = CreateDocumentFilePath(new[] { "B", "Foo" }, "File2.cs");   // file2 is in <root>\B\Foo\
+
+            var code =
+$@"
+<Workspace>
+    <Project Language=""C#"" AssemblyName=""Assembly1"" FilePath=""{ProjectFilePath}"" DefaultNamespace=""{defaultNamespace}"" CommonReferences=""true"">
+        <Document Folders=""{documentPath1.folder}"" FilePath=""{documentPath1.filePath}""> 
+namespace [||]{declaredNamespace}
+{{    
+    class Class1
+    {{
+    }}
+}}  
+        </Document>        
+        <Document Folders=""{documentPath2.folder}"" FilePath=""{documentPath2.filePath}""> 
+namespace Foo
+{{    
+    class Class2
+    {{
+    }}
+}}  
+        </Document>  
+    </Project>
+</Workspace>";
+            await TestMoveFileToMatchNamespace(code, expectedFolders);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CodeActionsSyncNamespace)]
+        public async Task MoveFile_FromOneFolderToAnother2()
+        {
+            var defaultNamespace = "A";
```

Comment:
some question on experience.

if a file moved to different folder is last file in the folder, will it delete the folder? also, would it be possible to do undo test in some way? I know undo is for next iteration. but just want to mention that undo experience seems one of most important thing for this feature.

Reply:
1. No, the empty folder will be left behind.
2. Not sure how to test undo yet, but I agree it's an important part that's missing currently. I looked at some similar refactoring (i.e. move type to new file), but couldn't find any undo test there. Will keep digging.